File: backend/internal/rules/transform.go
package rules

// WAFv2 TextTransformations, ported from src/lib/server/ruletransform.ts.
// Every type AWS accepts is handled except MD5, whose output is raw binary no
// pasted SearchString can express — that one is reported by name.

import (
	"encoding/base64"
	"encoding/hex"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type TransformResult struct {
	OK    bool
	Value string
	Type  string // the transform that stopped evaluation, when !OK
}

var htmlEntities = map[string]string{
	"&lt;": "<", "&gt;": ">", "&amp;": "&", "&quot;": `"`, "&apos;": "'", "&nbsp;": " ",
}

func codePoint(n int64) (string, bool) {
	if n < 0 || n > 0x10ffff {
		return "", false
	}
	return string(rune(n)), true
}

var (
	namedEntityRe = regexp.MustCompile(`(?i)&(?:lt|gt|amp|quot|apos|nbsp);`)
	decEntityRe   = regexp.MustCompile(`&#([0-9]+);?`)
	hexEntityRe   = regexp.MustCompile(`(?i)&#x([0-9a-f]+);?`)
)

func htmlEntityDecode(s string) string {
	s = namedEntityRe.ReplaceAllStringFunc(s, func(m string) string {
		if v, ok := htmlEntities[strings.ToLower(m)]; ok {
			return v
		}
		return m
	})
	s = decEntityRe.ReplaceAllStringFunc(s, func(m string) string {
		d := decEntityRe.FindStringSubmatch(m)[1]
		n, err := strconv.ParseInt(d, 10, 64)
		if err != nil {
			return m
		}
		if v, ok := codePoint(n); ok {
			return v
		}
		return m
	})
	return hexEntityRe.ReplaceAllStringFunc(s, func(m string) string {
		h := hexEntityRe.FindStringSubmatch(m)[1]
		n, err := strconv.ParseInt(h, 16, 64)
		if err != nil {
			return m
		}
		if v, ok := codePoint(n); ok {
			return v
		}
		return m
	})
}

func urlDecode(s string) string {
	// A malformed percent-escape decodes to itself in WAF; keep the raw value.
	if d, err := url.QueryUnescape(s); err == nil {
		return d
	}
	return s
}

var pctURe = regexp.MustCompile(`(?i)%u([0-9a-f]{4})`)

func urlDecodeUni(s string) string {
	return pctURe.ReplaceAllStringFunc(urlDecode(s), func(m string) string {
		n, _ := strconv.ParseInt(pctURe.FindStringSubmatch(m)[1], 16, 64)
		if v, ok := codePoint(n); ok {
			return v
		}
		return m
	})
}

func normalizePathT(s string) string {
	segs := []string{}
	for _, seg := range strings.Split(s, "/") {
		switch seg {
		case "", ".":
		case "..":
			if len(segs) > 0 {
				segs = segs[:len(segs)-1]
			}
		default:
			segs = append(segs, seg)
		}
	}
	trailing := ""
	if len(s) > 1 && strings.HasSuffix(s, "/") && len(segs) > 0 {
		trailing = "/"
	}
	return "/" + strings.Join(segs, "/") + trailing
}

var b64StrictRe = regexp.MustCompile(`^[A-Za-z0-9+/]+={0,2}$`)

func base64DecodeT(s string, lenient bool) string {
	var t string
	if lenient {
		var b strings.Builder
		for _, r := range s {
			if r >= 'A' && r <= 'Z' || r >= 'a' && r <= 'z' || r >= '0' && r <= '9' ||
				r == '+' || r == '/' || r == '=' {
				b.WriteRune(r)
			}
		}
		t = b.String()
		if t == "" {
			return s
		}
	} else {
		t = strings.NewReplacer("\r", "", "\n", "").Replace(s)
		// WAF leaves an undecodable value alone.
		if t == "" || len(t)%4 != 0 || !b64StrictRe.MatchString(t) {
			return s
		}
	}
	pad := strings.TrimRight(t, "=")
	if d, err := base64.RawStdEncoding.DecodeString(pad); err == nil {
		return string(d)
	}
	return s
}

func hexDecodeT(s string) string {
	t := strings.TrimSpace(s)
	if t == "" || len(t)%2 != 0 {
		return s
	}
	if d, err := hex.DecodeString(t); err == nil {
		return string(d)
	}
	return s
}

var sqlHexRe = regexp.MustCompile(`\b0x((?:[0-9a-fA-F]{2})+)\b`)

// SQL hex literals: 0x646f67 -> dog
func sqlHexDecode(s string) string {
	return sqlHexRe.ReplaceAllStringFunc(s, func(m string) string {
		h := sqlHexRe.FindStringSubmatch(m)[1]
		if d, err := hex.DecodeString(h); err == nil {
			return string(d)
		}
		return m
	})
}

var escapeChars = map[string]string{
	"a": "\x07", "b": "\b", "f": "\f", "n": "\n", "r": "\r", "t": "\t", "v": "\v",
	`\`: `\`, "?": "?", "'": "'", `"`: `"`,
}

var (
	escURe    = regexp.MustCompile(`\\u([0-9a-fA-F]{4})`)
	escXRe    = regexp.MustCompile(`\\x([0-9a-fA-F]{2})`)
	escORe    = regexp.MustCompile(`\\([0-7]{1,3})`)
	escCRe    = regexp.MustCompile(`\\([abfnrtv\\?'"])`)
	commentRe = regexp.MustCompile(`(?s)/\*.*?(\*/|$)`)
	wsRe      = regexp.MustCompile(`\s+`)
)

func replCP(re *regexp.Regexp, base int) func(string) string {
	return func(m string) string {
		n, _ := strconv.ParseInt(re.FindStringSubmatch(m)[1], base, 64)
		if v, ok := codePoint(n); ok {
			return v
		}
		return m
	}
}

func escapeSeqDecode(s string) string {
	s = escURe.ReplaceAllStringFunc(s, replCP(escURe, 16))
	s = escXRe.ReplaceAllStringFunc(s, replCP(escXRe, 16))
	s = escORe.ReplaceAllStringFunc(s, replCP(escORe, 8))
	return escCRe.ReplaceAllStringFunc(s, func(m string) string {
		c := escCRe.FindStringSubmatch(m)[1]
		if v, ok := escapeChars[c]; ok {
			return v
		}
		return m
	})
}

var (
	cssContRe = regexp.MustCompile(`\\\r?\n`)
	cssHexRe  = regexp.MustCompile(`\\([0-9a-fA-F]{1,6})[ \t]?`)
	cssAnyRe  = regexp.MustCompile(`\\([^\r\n])`)
)

// CSS escapes: \3c -> "<", "\<newline>" is a line continuation, "\x" -> "x".
func cssDecode(s string) string {
	s = cssContRe.ReplaceAllString(s, "")
	s = cssHexRe.ReplaceAllStringFunc(s, replCP(cssHexRe, 16))
	return cssAnyRe.ReplaceAllString(s, "$1")
}

var (
	jsBraceRe = regexp.MustCompile(`\\u\{([0-9a-fA-F]{1,6})\}`)
	jsCRe     = regexp.MustCompile(`\\([bfnrtv0'"\\/])`)
)

func jsDecode(s string) string {
	s = jsBraceRe.ReplaceAllStringFunc(s, replCP(jsBraceRe, 16))
	s = escURe.ReplaceAllStringFunc(s, replCP(escURe, 16))
	s = escXRe.ReplaceAllStringFunc(s, replCP(escXRe, 16))
	s = escORe.ReplaceAllStringFunc(s, replCP(escORe, 8))
	return jsCRe.ReplaceAllStringFunc(s, func(m string) string {
		c := jsCRe.FindStringSubmatch(m)[1]
		if c == "0" {
			return "\x00"
		}
		if v, ok := escapeChars[c]; ok {
			return v
		}
		return c
	})
}

func utf8ToUnicode(s string) string {
	var out strings.Builder
	for _, ch := range s {
		if ch > 0x7f {
			out.WriteString("%u")
			h := strconv.FormatInt(int64(ch), 16)
			for len(h) < 4 {
				h = "0" + h
			}
			out.WriteString(h)
		} else {
			out.WriteRune(ch)
		}
	}
	return out.String()
}

// AWS CMD_LINE: drop \ " ' ^, collapse whitespace to one space, lowercase.
func cmdLine(s string) string {
	s = strings.NewReplacer(`\`, "", `"`, "", "'", "", "^", "").Replace(s)
	return strings.ToLower(strings.TrimSpace(wsRe.ReplaceAllString(s, " ")))
}

func transformOne(value, typ string) (string, bool) {
	switch typ {
	case "NONE":
		return value, true
	case "LOWERCASE":
		return strings.ToLower(value), true
	case "TRIM":
		return strings.TrimSpace(value), true
	case "COMPRESS_WHITE_SPACE":
		return wsRe.ReplaceAllString(value, " "), true
	case "REMOVE_NULLS":
		return strings.ReplaceAll(value, "\x00", ""), true
	case "REPLACE_NULLS":
		return strings.ReplaceAll(value, "\x00", " "), true
	case "URL_DECODE":
		return urlDecode(value), true
	case "URL_DECODE_UNI":
		return urlDecodeUni(value), true
	case "HTML_ENTITY_DECODE":
		return htmlEntityDecode(value), true
	case "BASE64_DECODE":
		return base64DecodeT(value, false), true
	case "BASE64_DECODE_EXT":
		return base64DecodeT(value, true), true
	case "HEX_DECODE":
		return hexDecodeT(value), true
	case "SQL_HEX_DECODE":
		return sqlHexDecode(value), true
	case "REPLACE_COMMENTS":
		return commentRe.ReplaceAllString(value, " "), true
	case "ESCAPE_SEQ_DECODE":
		return escapeSeqDecode(value), true
	case "CSS_DECODE":
		return cssDecode(value), true
	case "JS_DECODE":
		return jsDecode(value), true
	case "UTF8_TO_UNICODE":
		return utf8ToUnicode(value), true
	case "NORMALIZE_PATH":
		return normalizePathT(value), true
	case "NORMALIZE_PATH_WIN":
		return normalizePathT(strings.ReplaceAll(value, `\`, "/")), true
	case "CMD_LINE":
		return cmdLine(value), true
	default:
		// MD5 lands here on purpose: its output is binary, so no pasted
		// SearchString could be compared against it honestly.
		return "", false
	}
}

func ApplyTransforms(value string, transforms any) TransformResult {
	list, _ := transforms.([]any)
	type tf struct {
		priority float64
		typ      string
	}
	ordered := []tf{}
	for _, t := range list {
		rec, ok := t.(map[string]any)
		if !ok {
			continue
		}
		p, _ := rec["Priority"].(float64)
		typ, _ := rec["Type"].(string)
		ordered = append(ordered, tf{p, typ})
	}
	sort.SliceStable(ordered, func(i, j int) bool { return ordered[i].priority < ordered[j].priority })

	out := value
	for _, t := range ordered {
		next, ok := transformOne(out, t.typ)
		if !ok {
			typ := t.typ
			if typ == "" {
				typ = "(이름 없는 변환)"
			}
			return TransformResult{OK: false, Type: typ}
		}
		out = next
	}
	return TransformResult{OK: true, Value: out}
}

File: backend/rules/transform.ts
// WAFv2 TextTransformations. Every type AWS accepts is handled except MD5, whose
// output is raw binary no pasted SearchString can express — that one is reported
// by name.

// NORMALIZE_PATH is not implemented here. The assembler builds its path patterns
// with config/paths.ts and this simulator evaluates them, so a second copy of
// that normalisation is a copy that can drift — and it had: this file kept the
// trailing slash while config/paths.ts dropped it, meaning the two halves of the
// same screen could describe different strings. One definition, imported.
import { normalizePathSegments } from "../config/paths.ts";

export interface TransformResult {
  ok: boolean;
  value: string;
  /** The transform that stopped evaluation, when !ok. */
  type: string;
}

const HTML_ENTITIES: Record<string, string> = {
  "&lt;": "<",
  "&gt;": ">",
  "&amp;": "&",
  "&quot;": '"',
  "&apos;": "'",
  "&nbsp;": " ",
};

function codePoint(n: number): string | null {
  if (!Number.isFinite(n) || n < 0 || n > 0x10ffff) return null;
  try {
    return String.fromCodePoint(n);
  } catch {
    return null;
  }
}

function htmlEntityDecode(s: string): string {
  return s
    .replace(/&(?:lt|gt|amp|quot|apos|nbsp);/gi, (m) => HTML_ENTITIES[m.toLowerCase()] ?? m)
    .replace(/&#([0-9]+);?/g, (m, d: string) => codePoint(Number.parseInt(d, 10)) ?? m)
    .replace(/&#x([0-9a-f]+);?/gi, (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m);
}

function urlDecode(s: string): string {
  // A malformed percent-escape decodes to itself in WAF; keep the raw value.
  try {
    return decodeURIComponent(s.replace(/\+/g, " "));
  } catch {
    return s;
  }
}

function urlDecodeUni(s: string): string {
  return urlDecode(s).replace(
    /%u([0-9a-f]{4})/gi,
    (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m,
  );
}

const B64_STRICT_RE = /^[A-Za-z0-9+/]+={0,2}$/;

function base64DecodeTransform(s: string, lenient: boolean): string {
  let t: string;
  if (lenient) {
    t = s.replace(/[^A-Za-z0-9+/=]/g, "");
    if (t === "") return s;
  } else {
    t = s.replace(/[\r\n]/g, "");
    // WAF leaves an undecodable value alone.
    if (t === "" || t.length % 4 !== 0 || !B64_STRICT_RE.test(t)) return s;
  }
  try {
    return Buffer.from(t.replace(/=+$/, ""), "base64").toString("utf8");
  } catch {
    return s;
  }
}

function hexDecodeTransform(s: string): string {
  const t = s.trim();
  if (t === "" || t.length % 2 !== 0 || !/^[0-9a-fA-F]+$/.test(t)) return s;
  return Buffer.from(t, "hex").toString("utf8");
}

/** SQL hex literals: 0x646f67 -> dog */
function sqlHexDecode(s: string): string {
  return s.replace(/\b0x((?:[0-9a-fA-F]{2})+)\b/g, (m, h: string) =>
    Buffer.from(h, "hex").toString("utf8") || m,
  );
}

const ESCAPE_CHARS: Record<string, string> = {
  a: "\x07",
  b: "\b",
  f: "\f",
  n: "\n",
  r: "\r",
  t: "\t",
  v: "\v",
  "\\": "\\",
  "?": "?",
  "'": "'",
  '"': '"',
};

function escapeSeqDecode(s: string): string {
  return s
    .replace(/\\u([0-9a-fA-F]{4})/g, (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m)
    .replace(/\\x([0-9a-fA-F]{2})/g, (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m)
    .replace(/\\([0-7]{1,3})/g, (m, o: string) => codePoint(Number.parseInt(o, 8)) ?? m)
    .replace(/\\([abfnrtv\\?'"])/g, (m, c: string) => ESCAPE_CHARS[c] ?? m);
}

/** CSS escapes: \3c -> "<", "\<newline>" is a line continuation, "\x" -> "x". */
function cssDecode(s: string): string {
  return s
    .replace(/\\\r?\n/g, "")
    .replace(/\\([0-9a-fA-F]{1,6})[ \t]?/g, (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m)
    .replace(/\\([^\r\n])/g, "$1");
}

function jsDecode(s: string): string {
  return s
    .replace(/\\u\{([0-9a-fA-F]{1,6})\}/g, (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m)
    .replace(/\\u([0-9a-fA-F]{4})/g, (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m)
    .replace(/\\x([0-9a-fA-F]{2})/g, (m, h: string) => codePoint(Number.parseInt(h, 16)) ?? m)
    .replace(/\\([0-7]{1,3})/g, (m, o: string) => codePoint(Number.parseInt(o, 8)) ?? m)
    .replace(/\\([bfnrtv0'"\\/])/g, (_m, c: string) =>
      c === "0" ? "\x00" : (ESCAPE_CHARS[c] ?? c),
    );
}

function utf8ToUnicode(s: string): string {
  let out = "";
  for (const ch of s) {
    const cp = ch.codePointAt(0)!;
    if (cp > 0x7f) out += "%u" + cp.toString(16).padStart(4, "0");
    else out += ch;
  }
  return out;
}

/** AWS CMD_LINE: drop \ " ' ^, collapse whitespace to one space, lowercase. */
function cmdLine(s: string): string {
  return s
    .replace(/[\\"'^]/g, "")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
}

function transformOne(value: string, type: string): string | null {
  switch (type) {
    case "NONE":
      return value;
    case "LOWERCASE":
      return value.toLowerCase();
    case "TRIM":
      return value.trim();
    case "COMPRESS_WHITE_SPACE":
      return value.replace(/\s+/g, " ");
    case "REMOVE_NULLS":
      return value.replaceAll("\x00", "");
    case "REPLACE_NULLS":
      return value.replaceAll("\x00", " ");
    case "URL_DECODE":
      return urlDecode(value);
    case "URL_DECODE_UNI":
      return urlDecodeUni(value);
    case "HTML_ENTITY_DECODE":
      return htmlEntityDecode(value);
    case "BASE64_DECODE":
      return base64DecodeTransform(value, false);
    case "BASE64_DECODE_EXT":
      return base64DecodeTransform(value, true);
    case "HEX_DECODE":
      return hexDecodeTransform(value);
    case "SQL_HEX_DECODE":
      return sqlHexDecode(value);
    case "REPLACE_COMMENTS":
      return value.replace(/\/\*[\s\S]*?(\*\/|$)/g, " ");
    case "ESCAPE_SEQ_DECODE":
      return escapeSeqDecode(value);
    case "CSS_DECODE":
      return cssDecode(value);
    case "JS_DECODE":
      return jsDecode(value);
    case "UTF8_TO_UNICODE":
      return utf8ToUnicode(value);
    case "NORMALIZE_PATH":
      return normalizePathSegments(value);
    case "NORMALIZE_PATH_WIN":
      return normalizePathSegments(value.replaceAll("\\", "/"));
    case "CMD_LINE":
      return cmdLine(value);
    default:
      // MD5 lands here on purpose: its output is binary, so no pasted
      // SearchString could be compared against it honestly.
      return null;
  }
}

export function applyTransforms(value: string, transforms: unknown): TransformResult {
  const list = Array.isArray(transforms) ? transforms : [];
  const ordered = list
    .filter((t): t is Record<string, unknown> => t !== null && typeof t === "object")
    .map((t) => ({
      priority: typeof t.Priority === "number" ? t.Priority : 0,
      type: typeof t.Type === "string" ? t.Type : "",
    }));
  // Array#sort is stable, so equal priorities keep their declared order.
  ordered.sort((a, b) => a.priority - b.priority);

  let out = value;
  for (const t of ordered) {
    const next = transformOne(out, t.type);
    if (next === null) {
      return { ok: false, value: "", type: t.type || "(이름 없는 변환)" };
    }
    out = next;
  }
  return { ok: true, value: out, type: "" };
}
